cocos/2d/CCTMXLayer.h

```diff
@@ -353,6 +372,65 @@ class CC_DLL TMXLayer : public SpriteBatchNode
     int _hexSideLength;
     /** properties from the layer. They can be added using Tiled */
     ValueMap _properties;
+
+    /** map from gid of animated tile to its instance. Also useful for optimization*/
+    std::map<uint32_t, std::vector<Vec2>> _animTileCoord;
+    /** pointer to the tile animation manager of this layer */
+    TMXTileAnimManager *_tileAnimManager = nullptr;
+};
+
+/** @brief TMXTileAnimTask represents the frame-tick task of an animated tile.
+ * It is a assistant class for TMXTileAnimTicker.
+ */
+class TMXTileAnimTask
+{
+public:
+    TMXTileAnimTask(TMXLayer *layer, TMXTileAnimInfo *animation, Vec2 &tilePos);
+    /** start the animation task */
+    void start();
+    /** stop the animation task */
+    void stop();
+    bool isRunning() {
+        return _isRunning;
+    }
+
+protected:
+    /** set texture of tile to current frame */
+    void setCurrFrame();
+    /** tick to next frame and schedule next tick */
+    void tickAndScheduleNext(float dt);
+
+    bool _isRunning;
+    /** key of schedule task for specific animated tile */
+    std::string _key;
+    TMXLayer *_layer;
+    /** position of the animated tile */
+    Vec2 _tilePosition;
+    /** AnimationInfo on this tile */
+    TMXTileAnimInfo *_animation;
+    /** Index of the frame that should be drawn currently */
+    uint32_t _currentFrame;
+    uint32_t _frameCount;
+};
+
+/** @brief TMXTileAnimManager controls all tile animation of a layer.
+ */
+class TMXTileAnimManager : public Ref
+{
+public:
+    static TMXTileAnimManager * create(TMXLayer *layer);
+    explicit TMXTileAnimManager(TMXLayer *layer);
+
+    /** start all tile animations */
+    void startAll();
+    /** stop all tile animations */
+    void stopAll();
+
+
+    bool _started = false;
+    /** vector contains all tasks of this layer */
+    std::vector<TMXTileAnimTask> _tasks;
+    TMXLayer* _layer;
```

Comment:
Why variables be public?

Reply:
Got it, but about `_tasks`, if users want to control a particular tile animation, then we'll need a getter for it, I wonder if the lua value conversion error will appear again?

Reply:
But now how to access `_tasks` in lua?

Reply:
Hmmm, you are right

Reply:
fixed. Turned to cocos2d::Vector.